### PR TITLE
[macOS] Isolate openGL rendering to FlutterOpenGLRenderer

### PR DESCRIPTION
**Pull request by iskakaushik**
This is in preparation for having an equivalent FlutterMetalRenderer

### CODE REVIEW

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  [_textures removeObjectForKey:@(textureID)];
+}
+
+#pragma mark - Helper methods to create rendering config for embedder.
+
+- (FlutterRendererConfig)createRendererConfig {
+  const FlutterRendererConfig rendererConfig = {
+      .type = kOpenGL,
+      .open_gl.struct_size = sizeof(FlutterOpenGLRendererConfig),
+      .open_gl.make_current = (BoolCallback)OnMakeCurrent,
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:154]: Nit: Here and elsewhere, C-style casts in ObjC++ code.

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  FlutterExternalTextureGL* FlutterTexture =
+      [[FlutterExternalTextureGL alloc] initWithFlutterTexture:texture];
+  int64_t textureID = [FlutterTexture textureID];
+  _embedderAPI.RegisterExternalTexture(_engine, textureID);
+  _textures[@(textureID)] = FlutterTexture;
+  return textureID;
+}
+
+- (void)textureFrameAvailable:(int64_t)textureID {
+  _embedderAPI.MarkExternalTextureFrameAvailable(_engine, textureID);
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:140]: Nit: Return codes are ignored. Maybe log it at the very least.

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+- (BOOL)populateTextureWithIdentifier:(int64_t)textureID
+                        openGLTexture:(FlutterOpenGLTexture*)openGLTexture {
+  return [_textures[@(textureID)] populateTexture:openGLTexture];
+}
+
+- (int64_t)registerTexture:(id<FlutterTexture>)texture {
+  FlutterExternalTextureGL* FlutterTexture =
+      [[FlutterExternalTextureGL alloc] initWithFlutterTexture:texture];
+  int64_t textureID = [FlutterTexture textureID];
+  _embedderAPI.RegisterExternalTexture(_engine, textureID);
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:134]: Add the the registered textures only if you can successfully register the texture with the engine.

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  return [_flutterView frameBufferIDForSize:size];
+}
+
+- (NSOpenGLContext*)resourceContext {
+  if (!_resourceContext) {
+    NSOpenGLPixelFormatAttribute attributes[] = {
+        NSOpenGLPFAColorSize, 24, NSOpenGLPFAAlphaSize, 8, 0,
+    };
+    NSOpenGLPixelFormat* pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:attributes];
+    _resourceContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:109]: Just for my clarification: How is the `shareContext` `nil` here? Are we using creating this context first and then using it as the prototype for the main context?

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  NSMutableDictionary<NSNumber*, FlutterExternalTextureGL*>* _textures;
+
+  FlutterEngineProcTable _embedderAPI;
+}
+
+- (instancetype)initWithFlutterEngine:(FLUTTER_API_SYMBOL(FlutterEngine))engine {
+  self = [super init];
+  if (self) {
+    _engine = engine;
+    _textures = [[NSMutableDictionary alloc] init];
```
Comment (chinmaygarde) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:67]: Can you confirm that these TUs use ARC?

**Review by chinmaygarde — COMMENTED** [review]
I may have reviewed some code that you only moved. But LGTM otherwise.

**Review by chinmaygarde — APPROVED** [review]

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm (hunk 4 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:109]: Yup, exactly: https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm#L247

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm (hunk 1 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:154]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm (hunk 2 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:140]: Done!

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm (hunk 3 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:134]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm (hunk 5 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm:67]: Yup, they use ARC: https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/BUILD.gn#L88

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 6]
```diff
@@ -242,17 +174,8 @@ - (BOOL)runWithEntrypoint:(NSString*)entrypoint {
...
-      .open_gl.struct_size = sizeof(FlutterOpenGLRendererConfig),
-      .open_gl.make_current = (BoolCallback)OnMakeCurrent,
-      .open_gl.clear_current = (BoolCallback)OnClearCurrent,
-      .open_gl.present = (BoolCallback)OnPresent,
-      .open_gl.fbo_with_frame_info_callback = (UIntFrameInfoCallback)OnFBO,
-      .open_gl.fbo_reset_after_present = true,
-      .open_gl.make_resource_current = (BoolCallback)OnMakeResourceCurrent,
-      .open_gl.gl_external_texture_frame_callback = (TextureFrameCallback)OnAcquireExternalTexture,
-  };
+  [_openGLRenderer attachToFlutterView:_viewController.flutterView];
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:177]: There is no guarantee that there will be a view here. How are you handling later changes to the view controller given that you removed `setViewController:`?

File: shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Called by the engine when the context's buffers should be swapped.
+ */
+- (bool)present;
+
+/**
+ * Called by the engine when framebuffer object ID is requested.
+ */
+- (uint32_t)getFBO:(const FlutterFrameInfo*)info;
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.h:54]: "get" is not idiomatic ObjC naming. This should be called `identifierForFBO:` or similar.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h [hunk 8]
```diff
@@ -21,7 +21,7 @@
  * Coordinates a single instance of execution of a Flutter engine.
  */
 FLUTTER_EXPORT
-@interface FlutterEngine : NSObject <FlutterTextureRegistry, FlutterPluginRegistry>
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:24]: Did you follow the breaking change process for this change? You've changed the public API of a public class.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm (hunk 6 above)
Comment (iskakaushik) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:177]: Good point, will account for it.

**Pull request merged**